src/Dropzone/index.jsx

```diff
@@ -0,0 +1,282 @@
+}) => {
+  const [isMultipleDragged, setIsMultipleDragged] = useState(false);
+  const [errors, setErrors] = useState([]);
+  const [progress, setProgress] = useState(0);
+  const [fileName, setFileName] = useState(undefined);
+  const [controller, setController] = useState(undefined);
+
+  const {
+    onUploadProgress = () => {},
+    onUploadSuccess = () => {},
+    onUploadCancel = () => {},
+    onUploadError = () => {},
+  } = uploadCallbacks;
+  const {
+    uploadError: uploadErrorMsg,
+    invalidSize: invalidSizeMsg,
+    invalidType: invalidTypeMsg,
+    multipleDragged: multipleDraggedMsg,
+  } = errorMessages;
+
+  const onDragEnter = useCallback(async (e) => {
+    if (errors) {
+      setErrors([]);
+    }
+    const files = await fromEvent(e);
+    if (files && files.length > 1) {
+      setIsMultipleDragged(true);
+    }
+  }, [errors]);
+
+  const onDragLeave = useCallback(() => {
+    if (isMultipleDragged) {
+      setIsMultipleDragged(false);
+    }
+  }, [isMultipleDragged]);
+
+  const onDropRejected = (files) => {
+    if (!isMultipleDragged) {
+      setErrors(files[0].errors.map(error => {
+        switch (error.code) {
+          case ErrorCode.FileTooLarge:
+            return invalidSizeMsg || error.message;
+          case ErrorCode.FileInvalidType:
+            return invalidTypeMsg || error.message;
+          default:
+            return error.message;
+        }
+      }));
+    } else {
+      setIsMultipleDragged(false);
+    }
+  };
+
+  const handleProgressUpload = (progressEvent) => {
+    const percentValue = Math.round((progressEvent.loaded * 100) / progressEvent.total);
+    setProgress(percentValue);
+    onUploadProgress(percentValue, progressEvent);
+  };
+
+  const processUpload = (client, uploadUrl, data) => {
+    const newController = new AbortController();
+    setController(newController);
+
+    const config = {
+      onUploadProgress: handleProgressUpload,
+      signal: newController.signal,
+    };
+
+    client.post(uploadUrl, data, config)
+      .then(response => onUploadSuccess(response))
+      .catch(error => {
+        // check if request has been canceled before treating the exception as an upload error
+        if (error.code !== 'ERR_CANCELED') {
+          setErrors([uploadErrorMsg]);
+          onUploadError(error);
+        }
+      });
+  };
+
+  const onDropAccepted = async (files) => {
+    if (isMultipleDragged) {
+      return;
+    }
+
+    const file = files[0];
+    if (validator) {
+      const customValidationError = await validator(file);
+      if (customValidationError) {
+        setErrors([customValidationError]);
+        return;
+      }
+    }
+
+    if (errors) {
+      setErrors([]);
+    }
+
+    const { uploadUrl, client } = axiosConfig || {};
+
+    if (!client || !uploadUrl) {
+      return;
+    }
+
+    const formData = new FormData();
+    formData.append('file', file);
+    setFileName(file.name);
+
+    processUpload(client, uploadUrl, formData);
```

Comment:
Yes, great idea, this is much more flexible for consumers! This way we can also remove `onUploadSuccess` and `onUploadError` callbacks from the component, since now consumers can handle these event in their upload handler function which they pass to `onProcessUpload` prop. Refactored it, thanks!